app/server/lib/OIDCConfig.ts

```diff
@@ -0,0 +1,192 @@
+ *   - GRIST_OIDC_IDP_SCOPES="openid email profile"
+ */
+
+import * as express from 'express';
+import {GristLoginSystem, GristServer} from './GristServer';
+import { Client, generators, Issuer, UserinfoResponse } from 'openid-client';
+import {Sessions} from './Sessions';
+import log from 'app/server/lib/log';
+import {Permit} from './Permit';
+
+const CALLBACK_URL = '/oauth2/callback';
+const LOGIN_URL = '/oauth2/login';
+
+const LOGIN_ACTION = 'oidc-login';
+const WAIT_IN_MINUTES = 20;
+
+export class OIDCConfig {
+  private _client: Client;
+
+  public constructor(private _gristServer: GristServer) {
+  }
+
+  public async initOIDC(): Promise<void> {
+    if (!process.env.GRIST_OIDC_SP_HOST) { throw new Error('initOIDC requires GRIST_OIDC_SP_HOST to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_ISSUER) { throw new Error('initOIDC requires GRIST_OIDC_IDP_ISSUER to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_ID) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_ID to be set');
+    }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_SECRET) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_SECRET to be set');
+    }
+    const spHost: string = process.env.GRIST_OIDC_SP_HOST;
+    const issuer = await Issuer.discover(process.env.GRIST_OIDC_IDP_ISSUER);
+    this._client = new issuer.Client({
+      client_id: process.env.GRIST_OIDC_IDP_CLIENT_ID,
+      client_secret: process.env.GRIST_OIDC_IDP_CLIENT_SECRET,
+      redirect_uris: [ new URL(CALLBACK_URL, spHost).href ],
+      response_types: ['code'],
+    });
+  }
+
+  public addEndpoints(app: express.Application, sessions: Sessions): void {
+    app.get(LOGIN_URL, this.handleLogin.bind(this));
+    app.get(CALLBACK_URL, this.handleCallback.bind(this, sessions));
+  }
+
+  public async handleCallback(sessions: Sessions, req: express.Request, res: express.Response): Promise<void> {
+
+    try {
+      const params = this._client.callbackParams(req);
+      const { state } = params;
+      if (!state) {
+        throw new Error('Login or logout failed to complete');
+      }
+      const codeVerifier = await this._retrieveCodeVerifierFromPermit(state);
+      const tokenSet = await this._client.callback(
+        new URL(CALLBACK_URL, process.env.GRIST_OIDC_SP_HOST).href,
+        params,
+        { state, code_verifier: codeVerifier }
+      );
+      const userInfo = await this._client.userinfo(tokenSet);
+      const profile = this._makeUserProfileFromUserInfo(userInfo);
+      const scopedSession = sessions.getOrCreateSessionFromRequest(req);
+      await scopedSession.operateOnScopedSession(req, async (user) => Object.assign(user, {
+        profile,
+        accessToken: tokenSet.access_token,
+        refreshToken: tokenSet.refresh_token,
+      }));
+      res.redirect('/');
+    } catch (err) {
+      log.error(`OIDC callback failed: ${err.message}`);
+      res.status(500).send(`Cannot connect to OIDC provider: ${err.message}`);
```

Comment:
I wonder if this message may be misleading -- it seems there could be other reasons for failure than failure to connect (e.g. expired permit)

Reply:
Yeah, I made the message more generic.